go-controller/pkg/ovn/controller/services/repair.go

```diff
@@ -67,13 +67,30 @@ func (r *repair) runBeforeSync() {
 		r.unsyncedServices.Insert(key)
 	}
 
+	// Find all templates.
```

Comment:
Does this mean that an ovndb with the previous code will not get converted to use templates?

I am not sure if this is needed, but I am also wondering is a rollback from new db to a previous implementation that did not use templates would tolerate rows with the new template.

Reply:
> Does this mean that an ovndb with the previous code will not get converted to use templates?
> 

It should be.  This part here is supposed to clean up stale templates.  For example, if at startup we still have in the NB templates corresponding to services that were deleted in the meantime.  It shouldn't be often but i guess this can happen as there's a window when ovnkube-master is down (e.g., it crashed or upgraded).

> I am not sure if this is needed, but I am also wondering if a rollback from new db to a previous implementation that did not use templates would tolerate rows with the new template.

At least in OCP we remove the DBs on downgrade IIRC.  So it should be fine.  In general, if we don't touch the database schema and file and just downgrade northd and ovnkube-master we will end up with regular load balancers being correctly created for the services.  The only problem is that template variable values will not be removed from the DB (older ovnkube-masters don't know how to do that).  But that's the case with other features at downgrade already (load balancer groups, copp).

Reply:
Ah, makes perfect sense. I will go over this code more carefully.

Reply:
Yes, I can see the conversion logic now. I was looking at the wrong place. 🤦  Nice!